### DISCUSSION TITLE
Pimcore performance with millions of Data Objects

### DISCUSSION THREAD

**Discussion by RuudvdvNB** [pimcore/pimcore#10907] [Q&A / Support]
## Description
One of our Pimcore (v6.9.4) projects contains almost 8 million (!) products. At this moment these products don't have a lot of attributes filled because most product data is fetched from third parties through API's.

The issue that we are facing at the moment with Pimcore is the fact that grids/folders are not loading because of multiple queries in Pimcore that take a while to execute. This results in loading times more than 30 seconds after which ExtJS / Pimcore Admin throws a communication failure. Besides the grids the quick search isn't loading in time either.

There is also an open issue https://github.com/pimcore/pimcore/issues/9727 which is actually the same issue. We also faced this and decided to give our client admin role so this query wouldn't be the issue anymore.

During the Pimcore inspire i discussed this with @dvesh3 after which he asked me to open up a discusion. Hopefully we could get some insights on how we can improve these queries and/or maybe there are some changed needed in the Pimcore core?

## Slow queries
After some debugging we found a few queries that are slow and causing the current Pimcore admin timeouts. Even tho we now have these queries there might be more that are causing timeouts.

### Grids
Route name: `pimcore_admin_dataobject_dataobject_getfolder`
Executed in: `\Pimcore\Model\DataObject\AbstractObject\Dao::getClasses`
Duration: 23.7s
```sql
SELECT o_classId FROM objects WHERE o_path LIKE ? AND o_type = 'object' GROUP BY o_classId
```

Route name: `pimcore_admin_dataobject_dataobject_gridproxy`
Executed in: `\Pimcore\Model\DataObject\Listing\Dao::getTotalCount`
Duration: 1m35s
```sql
SELECT COUNT(*) FROM object_PRODUCT WHERE (o_path = '/products' OR o_path LIKE '/products/%') AND  object_PRODUCT.o_type IN ('object','folder')
```

### Quick search
Route name: `pimcore_admin_searchadmin_search_quicksearch`
Executed in: `\Pimcore\Bundle\AdminBundle\Controller\Searchadmin\SearchController::quicksearchAction`
Duration: 31.3s
```sql
SELECT * FROM search_backend_data WHERE ((MATCH (data, properties) AGAINST (? IN BOOLEAN MODE)) AND type != ?) ORDER BY (MATCH (data, properties) AGAINST (? IN BOOLEAN MODE)) DESC LIMIT ?
```

## Database info
Server: MariaDB 10.4

| Table | Rows | Size |
| --- | --- | --- |
| dependencies | ~12.4M | 1.1GB |
| objects | ~7.9M | 2.8GB |
| object_collection_prices | ~4.4M | 1.6GB |
| object_localized_data_PRODUCT | 42.5M | 2.5GB |
| object_localized_query_PRODUCT_de | ~7.9M | 427MB |
| object_localized_query_PRODUCT_en | ~7.9M | 427MB |
| object_localized_query_PRODUCT_fr | ~7.9M | 427MB |
| object_localized_query_PRODUCT_nl | ~7.9M | 427MB |
| object_localized_query_PRODUCT_sv | ~7.9M | 427MB |
| object_metadata_PRODUCT | ~23.8M | 14.2GB |
| object_query_PRODUCT | ~7.9M | 1.9GB |
| object_relations_PRODUCT | ~15M | 2.0GB |
| object_store_PRODUCT | ~7.9M | 2.0GB |
| properties | ~8M | 2.7GB |
| search_backend_data | ~8M | 6.4GB |
| versions | ~3M | 5.4GB |

**Comment by brusch** [comment]
Should be fixed by #11856

**Comment by luka-fstack** [comment]
@brusch 

I would like to reopen this issue due to persistent performance problems in the Data Objects grid.

#### **Overview**

Our client’s Pimcore project (v11.2.7) contains:

- ~1.5 million objects (products)  
- ~70 attributes per object

The Data Objects grid view typically takes **around 20 seconds to load**. Additionally, several user actions trigger unnecessary reloads, resulting in delays and communication failures.

#### **Steps to Reproduce**

1. Open grid view (initial load on client's server takes ~28 seconds for 1.5 million objects for a non-admin user).  
2. Select Grid Options → objects reload.  
3. Add a new attribute to selected grid columns → objects reload again.  
4. Click **Apply** → another reload.  
5. If the user interacts before loading completes, a **communication failure** occurs (as described in https://github.com/orgs/pimcore/discussions/9803).

#### **Root Causes**

1. **Redundant Object Reloads**  
   Triggered by (unnecessary?) Ajax requests when modifying grid options.  
   File:  
   `vendor/pimcore/admin-ui-classic-bundle/public/js/pimcore/object/helpers/gridConfigDialog.js:168` (Ajax request to `gridProxyAction` in `vendor/pimcore/admin-ui-classic-bundle/src/Controller/Admin/DataObject/DataObjectController.php`

2. **Slow Database Query (`getTotalCount()`)**  
   Located in:  
   `vendor/pimcore/admin-ui-classic-bundle/src/Controller/Admin/DataObject/DataObjectActionsTrait.php:140`  
   More severe for non-admin users due to additional subqueries (see https://github.com/pimcore/pimcore/issues/6383).

#### **Impact**

- Unnecessary server load (multiple object reloads)  
- Poor user experience due to slow response times  
- Communication failures when users do not wait for loading to complete

**Comment by fashxp** [comment]
Hi @luka-fstack  
please note, that we did some additional improvements in the grid with platform-version 2024.4, see also https://docs.pimcore.com/platform/Platform_Version/Release_Notes/2024.4#admin-ui-classic-bundle

It includes pimcore/pimcore version 11.5 and pimcore/admin-ui-classic-bundle 1.7 - which actually includes the performance improvements, see also https://github.com/pimcore/admin-ui-classic-bundle/releases/tag/v1.7.0 and specifically https://github.com/pimcore/admin-ui-classic-bundle/pull/767. 

Maybe it is worth a try to upgrade to that versions (makes sense anyway as this is the LTS version), or at least apply the patches manually for testing. 

Please also understand, that we won't do any further improvements in that regards to the admin-ui-classic anymore. In studio, we follow a completely different approach and solved these performance issues...
